File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp

class Solution {
public:
    
    int dp[1001][1001] = {};
    
    int fun(string& s1, string& s2, int m, int n){
        
        if(m == 0 || n == 0){
            return 0;
        }
        
        
        if(dp[m - 1][n - 1] > 0){
            return dp[m - 1][n - 1];
        }
        
        if(s1[m - 1] == s2[n - 1]){
            return dp[m - 1][n - 1] = 1 + fun(s1, s2, m - 1, n - 1);
        }
        
        int t1 = fun(s1, s2, m - 1, n);
        int t2 = fun(s1, s2, m , n - 1);
        
        return dp[m - 1][n - 1] = max(t1, t2);
    }
    
    int longestCommonSubsequence(string text1, string text2) {
        //memset(dp, -1, sizeof(dp));
        return fun(text1, text2, text1.length(), text2.length());
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int n = text1.length(), m = text2.length();
        int[][] dp = new int[n][m];

        for(int i = 0; i < n; i++){
            Arrays.fill(dp[i], -1);
        }

        return helper(text1, text2, n - 1, m - 1, dp);
    }

    private int helper(String s1, String s2, int idx1, int idx2, int[][] dp){
        if(idx1 < 0 || idx2 < 0){
            return 0;
        }

        if(dp[idx1][idx2] != -1){
            return dp[idx1][idx2];
        }

        if(s1.charAt(idx1) == s2.charAt(idx2)){
            return 1 + helper(s1, s2, idx1 - 1, idx2 - 1, dp);
        }

        return Math.max(helper(s1, s2, idx1 - 1, idx2, dp), helper(s1, s2, idx1, idx2 - 1, dp));
    }
}